File: src/fetch.rs
use anyhow::Context;
use anyhow::Result;
use anyhow::anyhow;
use chrono::{Datelike, TimeZone, Utc};
use std::{collections::HashMap, env};

use log::{debug, info};
use serde_json::Value;
use sqlx::SqlitePool;

use crate::{
    database::{next_event_filtered, EventType},
    models::{
        CalendarEvents, CurrentConstructorStandings, CurrentDriverStandings, DriverList,
        EventTracker, F1APIDriverStanding, SessionInfo, SessionResults,
    },
};

const F1_API_ENDPOINT: &str = "https://api.formula1.com/v1/event-tracker";
const F1_SESSION_ENDPOINT: &str = "https://livetiming.formula1.com/static";
const ERGAST_API_ENDPOINT: &str = "https://ergast.com/api/f1";

pub async fn fetch_json<T: serde::de::DeserializeOwned>(
    url: &str,
    headers: Option<reqwest::header::HeaderMap>,
) -> Result<T> {
    let client = reqwest::Client::new();
    let mut request = client.get(url);

    if let Some(headers) = headers {
        request = request.headers(headers);
    }

    let body = request.send().await?.text().await?;
    let data: T = serde_json::from_str(body.trim_start_matches('\u{feff}'))?;

    Ok(data)
}

fn parse_datetime(
    date: &str,
    offset: &str,
) -> Result<chrono::DateTime<chrono::Utc>, chrono::ParseError> {
    let datetime_string = format!("{} {}", date, offset);
    info!("Parsing datetime: {}", datetime_string);
    chrono::DateTime::parse_from_str(&datetime_string, "%Y-%m-%dT%H:%M:%S %z")
        .map(|dt| dt.with_timezone(&chrono::Utc)) // Convert to Utc for universal comparison
}

// Fetch the next closest event from the F1 API
pub async fn fetch_next_event(pool: &SqlitePool) -> Result<Option<String>> {
    let data = fetch_json::<EventTracker>(
        F1_API_ENDPOINT,
        Some({
            let mut headers = reqwest::header::HeaderMap::new();
            headers.insert(
                "apiKey",
                env::var("F1_API_KEY")
                    .expect("F1_API_KEY environment variable not set")
                    .parse()?,
            );
            headers.insert("locale", reqwest::header::HeaderValue::from_static("en"));

            headers
        }),
    )
    .await?;

    let closest_event = data
        .season_context
        .timetables
        .into_iter()
        .filter(|event| event.state == "upcoming")
        .min_by(|a, b| {
            let start_time_a = match parse_datetime(&a.start_time, &a.gmt_offset) {
                Ok(time) => time,
                Err(e) => {
                    eprintln!("Failed to parse start time for event a: {}", e);
                    return std::cmp::Ordering::Equal;
                }
            };
            let start_time_b = match parse_datetime(&b.start_time, &b.gmt_offset) {
                Ok(time) => time,
                Err(e) => {
                    eprintln!("Failed to parse start time for event b: {}", e);
                    return std::cmp::Ordering::Equal;
                }
            };
            start_time_a.cmp(&start_time_b)
        });

    info!("Closest event: {:?}", closest_event);
    match closest_event {
        Some(event) => {
            let closest_event_start_time =
                match parse_datetime(&event.start_time, &event.gmt_offset) {
                    Ok(time) => time,
                    Err(e) => {
                        eprintln!("Failed to parse start time for event: {}", e);
                        return Ok(None);
                    }
                };

            let time_to_event: i64 = closest_event_start_time
                .signed_duration_since(chrono::Utc::now())
                .num_minutes();
            info!("Time to event: {} minutes", time_to_event);

            if time_to_event > 0 && time_to_event <= 5 {
                Ok(Some(format!(
                    "🏎️ \x02{}: {}\x02 begins in 5 minutes.",
                    data.race.meeting_official_name, event.description
                )))
            } else {
                Ok(None)
            }
        }
        None => match next_event_filtered(pool, None).await {
            Ok(event) => match event {
                Some((meeting_name, event_name, start_time)) => {
                    let event_time = Utc
                        .timestamp_opt(start_time, 0)
                        .single()
                        .ok_or("Invalid timestamp")
                        .unwrap();

                    let time_to_event: i64 = event_time
                        .signed_duration_since(chrono::Utc::now())
                        .num_minutes();
                    info!("Time to DB event: {} minutes", time_to_event);

                    if time_to_event > 0 && time_to_event <= 5 {
                        Ok(Some(format!(
                            "🏎️ \x02{}: {}\x02 begins in 5 minutes.",
                            meeting_name,
                            event_name.to_str()
                        )))
                    } else {
                        Ok(None)
                    }
                }
                None => Ok(None),
            },
            Err(_) => Ok(None),
        },
    }
}

pub async fn fetch_driver_list(path: &str, pool: &SqlitePool) -> Result<()> {
    let driver_list_path = format!("{}/{}DriverList.json", F1_SESSION_ENDPOINT, path);
    info!("Fetching driver list for {}", driver_list_path);
    let response_result: Result<HashMap<String, Value>, _> =
        fetch_json(driver_list_path.as_str(), None).await;

    match response_result {
        Ok(response) => {
            for driver_value in response.values() {
                let driver: DriverList = serde_json::from_value(driver_value.clone())?;
                sqlx::query!("INSERT INTO driver_list (racing_number, reference, first_name, last_name, full_name, broadcast_name, tla, country_code, team_name, team_color) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) ON CONFLICT (racing_number) DO UPDATE SET reference = ?, first_name = ?, last_name = ?, full_name = ?, broadcast_name = ?, tla = ?, country_code = ?, team_name = ?, team_color = ?",
                    driver.racing_number,
                    driver.reference,
                    driver.first_name,
                    driver.last_name,
                    driver.full_name,
                    driver.broadcast_name,
                    driver.tla,
                    driver.country_code,
                    driver.team_name,
                    driver.team_color,
                    driver.reference,
                    driver.first_name,
                    driver.last_name,
                    driver.full_name,
                    driver.broadcast_name,
                    driver.tla,
                    driver.country_code,
                    driver.team_name,
                    driver.team_color,
                ).execute(pool).await?;
            }
        }
        Err(e) => {
            eprintln!("Failed to fetch driver list: {}", e);
            return Ok(());
        }
    };

    Ok(())
}

#[derive(Debug)]
struct EssentialDriverData {
    position: String,
    racing_number: String,
    best_lap_time: String,
    interval_to_position_ahead: Option<String>,
}

async fn extract_position_and_number(
    data: HashMap<String, Value>,
    pool: &SqlitePool,
) -> Result<Vec<F1APIDriverStanding>> {
    let lines = data.get("Lines")
        .and_then(Value::as_object)
        .context("Failed to extract lines")?;

    let extracted_data: Vec<EssentialDriverData> = lines
        .iter()
        .map(|(_, driver_data)| {
            let position = driver_data["Position"].as_str().unwrap_or_default().to_string();
            let racing_number = driver_data["RacingNumber"].as_str().unwrap_or_default().to_string();
            let best_lap_time = driver_data
                .get("BestLapTime")
                .and_then(|v| v.get("Value"))
                .and_then(|v| v.as_str())
                .unwrap_or_default()
                .to_string();

            let interval_to_position_ahead = driver_data
                .get("Stats")
                .and_then(Value::as_array)
                .and_then(|stats| {
                    stats.iter().find_map(|stat| {
                        stat.get("TimeDifftoPositionAhead")
                            .and_then(Value::as_str)
                            .map(String::from)
                    })
                });

            EssentialDriverData {
                position,
                racing_number,
                best_lap_time,
                interval_to_position_ahead,
            }
        })
        .collect();

    let driver_list = sqlx::query!("SELECT tla, racing_number, team_name FROM driver_list")
        .fetch_all(pool)
        .await?;

    debug!("Driver list: {:?}", driver_list);
    debug!("Extracted data: {:?}", extracted_data);

    extracted_data
        .iter()
        .map(|data| {
            let driver = driver_list
                .iter()
                .find(|driver| {
                    driver.racing_number == data.racing_number.parse::<i64>().unwrap_or(-1)
                })
                .ok_or_else(|| anyhow!("Driver not found for racing number: {}", data.racing_number))?;

            let position = data.position.parse::<i8>()
                .map_err(|_| anyhow!("Invalid position: {}", data.position))?;

            Ok(F1APIDriverStanding {
                position,
                driver_name: driver.tla.clone(),
                team_name: driver.team_name.clone(),
                time: data.best_lap_time.clone(),
                difference: data.interval_to_position_ahead.clone(),
            })
        })
        .collect()
}

// Fetch results for a given path from the F1 API
pub async fn fetch_results(pool: &SqlitePool, path: &str) -> Result<String> {
    let result = sqlx::query!("SELECT data FROM results WHERE path = ?", path)
        .fetch_optional(pool)
        .await?;

    let session_result = match result {
        Some(previous_result) => {
            info!("Using cached results for {}", path);
            info!("Previous result: {}", previous_result.data);
            serde_json::from_str(&previous_result.data)
                .context("Failed to deserialize cached results")?
        }
        None => {
            debug!("Fetching TimingDataF1 for {}", path);
            let session = fetch_json::<HashMap<String, Value>>(
                format!("{}/{}TimingDataF1.json", F1_SESSION_ENDPOINT, path).as_str(),
                None,
            )
            .await
            .context("Failed to fetch TimingDataF1")?;

            debug!("Fetching SessionInfo for {}", path);
            let session_info: SessionInfo = fetch_json::<SessionInfo>(
                format!("{}/{}", F1_SESSION_ENDPOINT, "SessionInfo.json").as_str(),
                None,
            )
            .await
            .context("Failed to fetch SessionInfo")?;

            let standings = extract_position_and_number(session, pool)
                .await
                .context("Failed to extract position and number")?;

            let new_session_result = SessionResults {
                title: format!(
                    "{} {}",
                    session_info.meeting.official_name, session_info.name
                ),
                standings,
            };

            let json_result = serde_json::to_string(&new_session_result)
                .context("Failed to serialize new session result")?;
            sqlx::query!(
                "INSERT INTO results (path, data, event_id) VALUES (?, ?, (SELECT id FROM events WHERE start_time < unixepoch() AND event_type_id != 1 ORDER BY start_time DESC LIMIT 1))",
                path,
                json_result
            )
            .execute(pool)
            .await
            .context("Failed to insert new results into database")?;

            new_session_result
        }
    };

    let mut output = format!("🏎️ \x02{} Results\x02:", session_result.title);

    // Consider only the first 10 drivers sorted by the key "position" to avoid spamming
    let mut sorted_standings = session_result.standings;
    sorted_standings.sort_by_key(|standing| standing.position);

    for standing in sorted_standings.iter().take(10) {
        output.push_str(&format!(
            " {}. {} - \x0303[{}]\x03",
            standing.position, standing.driver_name, standing.time
        ));
    }

    Ok(output)
}

// Fetch the path of the current event from the F1 API
pub async fn read_current_event() -> Result<(String, bool)> {
    let session = fetch_json::<SessionInfo>(
        format!("{}/{}", F1_SESSION_ENDPOINT, "SessionInfo.json").as_str(),
        None,
    )
    .await
    .unwrap();

    Ok((session.path, session.archive_status.status == "Complete"))
}

pub async fn fetch_wdc_standings(pool: &SqlitePool) -> Result<CurrentDriverStandings> {
    match fetch_json::<CurrentDriverStandings>(
        format!("{}/current/driverStandings.json", ERGAST_API_ENDPOINT).as_str(),
        None,
    )
    .await {
        Ok(standings) => {
            let json_standings = serde_json::to_string(&standings)?;

            sqlx::query!(
                "INSERT INTO championship_standings (data, type) VALUES (?, 0) ON CONFLICT (type) DO UPDATE SET data = ?, create_time = CURRENT_TIMESTAMP",
                json_standings,
                json_standings
            )
            .execute(pool)
            .await?;

            Ok(standings)
        }
        Err(e) => {
            eprintln!("Failed to fetch WDC standings: {}", e);
            Err(e)
        }
    }
}

// Get current WDC standings
pub async fn return_wdc_standings(pool: &SqlitePool) -> Result<String> {
    let row = sqlx::query!("SELECT data FROM championship_standings WHERE type = 0 LIMIT 1",)
        .fetch_optional(pool)
        .await?;

    let standings: CurrentDriverStandings = match row {
        Some(row) => serde_json::from_str(&row.data).unwrap(),
        None => fetch_wdc_standings(pool).await?,
    };

    let mut output = format!(
        "🏆 \x02 FORMULA 1 {} WDC Standings\x02:",
        standings.mrdata.standings_table.season
    );

    standings
        .mrdata
        .standings_table
        .standings_lists
        .first()
        .unwrap()
        .driver_standings
        .iter()
        .take(10)
        .for_each(|standing| {
            output.push_str(&format!(
                " {}. {} - \x0303[{}]\x03",
                standing.position, standing.driver.code, standing.points
            ));
        });

    Ok(output)
}

pub async fn fetch_wcc_standings(pool: &SqlitePool) -> Result<CurrentConstructorStandings> {
    match fetch_json::<CurrentConstructorStandings>(
        format!("{}/current/constructorStandings.json", ERGAST_API_ENDPOINT).as_str(),
        None,
    )
    .await {
        Ok(standings) => {
            let json_standings = serde_json::to_string(&standings)?;

            sqlx::query!(
                "INSERT INTO championship_standings (data, type) VALUES (?, 1) ON CONFLICT (type) DO UPDATE SET data = ?, create_time = CURRENT_TIMESTAMP",
                json_standings,
                json_standings
            )
            .execute(pool)
            .await?;

            Ok(standings)
        }
        Err(e) => {
            eprintln!("Failed to fetch WCC standings: {}", e);
            Err(e)
        }
    }
}

// Get the current WCC standings
pub async fn return_wcc_standings(pool: &SqlitePool) -> Result<String> {
    let row = sqlx::query!("SELECT data FROM championship_standings WHERE type = 1 LIMIT 1")
        .fetch_optional(pool)
        .await?;

    let standings: CurrentConstructorStandings = match row {
        Some(row) => serde_json::from_str(&row.data).unwrap(),
        None => fetch_wcc_standings(pool).await?,
    };

    let mut output = format!(
        "🔧 \x02 FORMULA 1 {} WCC Standings\x02:",
        standings.mrdata.standings_table.season
    );

    standings
        .mrdata
        .standings_table
        .standings_lists
        .first()
        .unwrap()
        .constructor_standings
        .iter()
        .take(10)
        .for_each(|standing| {
            output.push_str(&format!(
                " {}. {} - \x0303[{}]\x03",
                standing.position, standing.constructor.name, standing.points
            ));
        });

    Ok(output)
}

// Fetch the current community F1 calendar
pub async fn refresh_current_calendar(pool: &SqlitePool, year: Option<i32>) -> Result<()> {
    let event_year = match year {
        Some(year) => year,
        None => Utc::now().year(),
    };

    let calendar_events = fetch_json::<CalendarEvents>(
        format!(
            "https://raw.githubusercontent.com/sportstimes/f1/main/_db/f1/{}.json",
            event_year
        )
        .as_str(),
        None,
    )
    .await
    .unwrap();

    for race in calendar_events.races {
        for (key, datetime_str) in race.sessions.iter() {
            if let Some(event_type) = EventType::from_str(key) {
                let datetime_str_with_millis =
                    if datetime_str.ends_with('Z') && !datetime_str.contains(".000Z") {
                        datetime_str.replace('Z', ".000Z")
                    } else {
                        datetime_str.to_string()
                    };

                match Utc.datetime_from_str(&datetime_str_with_millis, "%Y-%m-%dT%H:%M:%S%.3fZ") {
                    Ok(event_start_time) => {
                        let meeting_name = format!(
                            "{} FORMULA 1 {} GRAND PRIX {}",
                            event_type.to_emoji(),
                            race.name.to_uppercase(),
                            event_year
                        );
                        info!("{} -> {}", key, event_type.to_str());
                        info!("Inserting event: {}", meeting_name);

                        let slug = format!("{}-{}-{}", event_year, race.slug, key);
                        let event_type_id = event_type as i64;
                        let start_timestamp = event_start_time.timestamp();

                        sqlx::query!(
                            "INSERT INTO events (meeting_name, event_type_id, start_time, event_slug) VALUES (?, ?, ?, ?) ON CONFLICT (event_slug) DO UPDATE SET start_time = ?;",
                            meeting_name, 
                            event_type_id, 
                            start_timestamp, 
                            slug, 
                            start_timestamp
                        ).execute(pool).await?;
                    }
                    Err(error) => {
                        eprintln!("Can't parse date: {}: {}", &datetime_str, error)
                    }
                }
            } else {
                eprintln!("Unknown event type: {}", key);
            }
        }
    }

    Ok(())
}

File: src/fetch.ts
import {
	getDb,
	getEventTypeName,
	getNextEvent,
	storeChampionshipStandings,
	storeDrivers,
	storeEventResult,
} from "~/database";
import type {
	ConstructorMRData,
	Driver,
	DriverMRData,
	DriverStanding,
	SessionResults,
} from "~/types/models";
import { sessionKeyToEventType, stringToEventType } from "~/utils/events";

type CurrentConstructorStandings = { MRData: ConstructorMRData };
type CurrentDriverStandings = { MRData: DriverMRData };

const F1_SESSION_ENDPOINT = "https://livetiming.formula1.com/static";
const ERGAST_API_ENDPOINT = "https://api.jolpi.ca/ergast/f1";
const SESSION_KEY_ALIASES: Record<string, string> = {
	practice1: "fp1",
	practice2: "fp2",
	practice3: "fp3",
	sprintshootout: "sprintqualifying",
	sprintshootoutqualifying: "sprintqualifying",
};

async function fetchJson<T>(url: string): Promise<T> {
	const response = await fetch(url, {
		signal: AbortSignal.timeout(10000),
	});

	if (!response.ok) {
		throw new Error(`HTTP error! Status: ${response.status} for URL: ${url}`);
	}

	const text = await response.text();
	const cleanText = text.trim().replace(/^\uFEFF/, "");
	return JSON.parse(cleanText) as T;
}

export async function fetchDriverList(path: string): Promise<void> {
	console.log(`Fetching driver list for ${path}...`);

	try {
		const response = await fetchJson<
			Record<
				string,
				{
					RacingNumber: string;
					Reference: string;
					FirstName: string;
					LastName: string;
					FullName: string;
					BroadcastName: string;
					Tla: string;
					TeamName: string;
					TeamColour: string;
				}
			>
		>(`${F1_SESSION_ENDPOINT}/${path}DriverList.json`);

		const drivers: Driver[] = Object.values(response).map((driverData) => ({
			racingNumber: Number.parseInt(driverData.RacingNumber, 10),
			reference: driverData.Reference || "",
			firstName: driverData.FirstName || "",
			lastName: driverData.LastName || "",
			fullName: driverData.FullName || "",
			broadcastName: driverData.BroadcastName || "",
			tla: driverData.Tla || "",
			teamName: driverData.TeamName || "",
			teamColor: driverData.TeamColour || "#FFFFFF",
		}));

		await storeDrivers(drivers);
		console.log(`Successfully processed ${drivers.length} drivers`);
	} catch (error) {
		console.error("Error fetching or processing driver list:", error);
	}
}

export async function readCurrentEvent(): Promise<{
	path: string;
	isComplete: boolean;
}> {
	const response = await fetchJson<{
		ArchiveStatus: {
			Status: string;
		};
		Path: string;
	}>(`${F1_SESSION_ENDPOINT}/SessionInfo.json`);

	return {
		path: response.Path,
		isComplete: response.ArchiveStatus.Status === "Complete",
	};
}

interface TimingLine {
	Position?: string;
	RacingNumber?: string;
	BestLapTime?: {
		Value?: string;
	};
	Stats?: Array<{
		TimeDifftoPositionAhead?: string;
	}>;
}

interface SessionResultEntry {
	position: string;
	Driver: {
		code?: string;
	};
}

interface RaceTableResponse<T> {
	MRData: {
		RaceTable: {
			season: string;
			Races: T[];
		};
	};
}

export async function fetchResults(path: string): Promise<string> {
	console.log(`Fetching results for ${path}...`);
	const db = await getDb();

	try {
		const eventQuery = `
			SELECT e.meeting_name, et.name as event_type_name, r.data
			FROM results r
			JOIN events e ON r.event_id = e.id
			JOIN event_type et ON e.event_type_id = et.id
			WHERE r.path = ?
			LIMIT 1
		`;

		const result = await db.execute({
			sql: eventQuery,
			args: [path],
		});

		let sessionResult: SessionResults;

		if (result.rows.length > 0) {
			console.log(`Using cached results for ${path}`);
			try {
				const row = result.rows[0];
				if (typeof row.meeting_name !== "string" || typeof row.event_type_name !== "string") {
					throw new Error("Cached result metadata is invalid");
				}
				const data = JSON.parse(row.data as string) as SessionResults;
				sessionResult = {
					...data,
					title: `${row.meeting_name}: ${row.event_type_name}`,
				};
			} catch (parseError) {
				console.error("Error parsing cached results, fetching fresh data:", parseError);
				sessionResult = await fetchFreshResults(path);
			}
		} else {
			sessionResult = await fetchFreshResults(path);
		}

		let output = `🏎️ \x02${sessionResult.title} Results\x02:`;
		for (const standing of sessionResult.standings.slice(0, 10)) {
			output += ` ${standing.position}. ${standing.driverName} - \x0303[${standing.time}]\x03`;
		}
		return output;
	} catch (error) {
		console.error("Error fetching results:", error);
		return `Error fetching results: ${error instanceof Error ? error.message : "Unknown error"}`;
	}
}

async function fetchFreshResults(path: string): Promise<SessionResults> {
	console.log(`Fetching TimingDataF1 for ${path}...`);
	const timingData = await fetchJson<Record<string, unknown>>(
		`${F1_SESSION_ENDPOINT}/${path}TimingDataF1.json`,
	);

	console.log("Fetching SessionInfo...");
	const sessionInfoResponse = await fetchJson<{
		Meeting: {
			OfficialName: string;
		};
	}>(`${F1_SESSION_ENDPOINT}/SessionInfo.json`);

	const lines = timingData.Lines as Record<string, TimingLine>;
	if (!lines) {
		throw new Error("Failed to extract lines from timing data");
	}

	const driverList = await getDb().then((db) =>
		db.execute(`
			SELECT racing_number, tla, team_name 
			FROM driver_list
		`),
	);
	const driversMap = new Map(
		driverList.rows.map((row) => [row.racing_number as number, { tla: row.tla as string, team_name: row.team_name as string }]),
	);
	const standings: DriverStanding[] = [];
	for (const driverData of Object.values(lines)) {
		if (!driverData.Position || !driverData.RacingNumber) {
			continue;
		}

		const position = Number.parseInt(driverData.Position, 10);
		const racingNumber = Number.parseInt(driverData.RacingNumber, 10);
		const driver = driversMap.get(racingNumber);
		if (!driver) {
			console.warn(`Driver with racing number ${racingNumber} not found in database`);
			continue;
		}

		let difference: string | undefined;
		if (driverData.Stats && Array.isArray(driverData.Stats)) {
			difference = driverData.Stats.find((stat) => stat.TimeDifftoPositionAhead !== undefined)
				?.TimeDifftoPositionAhead;
		}

		standings.push({
			position,
			driverName: driver.tla,
			teamName: driver.team_name,
			time: driverData.BestLapTime?.Value || "",
			difference,
		});
	}
	standings.sort((a, b) => a.position - b.position);
	const lastSegment = path.replace(/\/+$/, "").split("/").filter(Boolean).at(-1) ?? "";
	const normalizedKey = lastSegment
		.split("_")
		.filter(Boolean)
		.filter((part, index) => !(index === 0 && /^\d{4}-\d{2}-\d{2}$/.test(part)))
		.join("")
		.toLowerCase();
	const sessionKey = SESSION_KEY_ALIASES[normalizedKey] || normalizedKey;

	let eventType = sessionKey ? sessionKeyToEventType(sessionKey) : null;
	if (eventType === null && sessionKey) {
		const fallback = stringToEventType(sessionKey);
		eventType = typeof fallback === "number" ? fallback : null;
	}
	const meetingName = sessionInfoResponse.Meeting.OfficialName;
	const sessionName = eventType !== null ? await getEventTypeName(eventType) : "";
	const sessionResult: SessionResults = {
		title: `${meetingName}${sessionName ? `: ${sessionName}` : ""}`,
		standings,
	};

	const db = await getDb();
	try {
		if (eventType === null) {
			console.warn("Unknown session type in path, skipping result storage for:", path);
		} else {
			const exact = await db.execute({
				sql: "SELECT id FROM events WHERE meeting_name = ? AND event_type_id = ? LIMIT 1",
				args: [meetingName, eventType],
			});
			const eventId =
				exact.rows[0]?.id ??
				(await db.execute({
					sql: "SELECT id FROM events WHERE meeting_name = ? LIMIT 1",
					args: [meetingName],
				})).rows[0]?.id;
			if (eventId) {
				await storeEventResult(eventId as number, path, sessionResult);
			} else {
				console.warn("Could not find matching event to store results for", meetingName, eventType);
			}
		}
	} catch (e) {
		console.error("Error while storing event result: ", e);
	}

	return sessionResult;
}

async function fetchStandings<T extends object>(
	type: number,
	url: string,
	formatFn: (data: T) => string,
): Promise<string | null> {
	console.log(`Fetching standings of type ${type}...`);
	const db = await getDb();

	try {
		const result = await db.execute({
			sql: "SELECT data FROM championship_standings WHERE type = ? LIMIT 1",
			args: [type],
		});

		let standings: T | undefined;
		if (result.rows.length > 0) {
			try {
				standings = JSON.parse(result.rows[0].data as string) as T;
			} catch (error) {
				console.error("Error parsing cached standings, fetching fresh data:", error);
			}
		}
		if (!standings) {
			const response = await fetchJson<{ MRData: unknown }>(url);
			standings = { MRData: response.MRData } as T;
			await storeChampionshipStandings(type, standings);
		}

		return formatFn(standings);
	} catch (error) {
		console.error(`Error fetching standings of type ${type}:`, error);
		return null;
	}
}

async function fetchCurrentStandings<T extends ConstructorMRData | DriverMRData>(
	type: number,
	label: "WCC" | "WDC",
	url: string,
): Promise<{ MRData: T } | null> {
	console.log(`Fetching ${label} standings...`);
	try {
		const standings = { MRData: (await fetchJson<{ MRData: T }>(url)).MRData };
		await storeChampionshipStandings(type, standings);
		return standings;
	} catch (error) {
		console.error(`Error fetching ${label} standings:`, error);
		return null;
	}
}

export async function fetchWccStandings(): Promise<CurrentConstructorStandings | null> {
	return fetchCurrentStandings(1, "WCC", `${ERGAST_API_ENDPOINT}/current/constructorstandings/?format=json`);
}

export async function fetchWdcStandings(): Promise<CurrentDriverStandings | null> {
	return fetchCurrentStandings(0, "WDC", `${ERGAST_API_ENDPOINT}/current/driverstandings/?format=json`);
}

export async function returnWccStandings(): Promise<string | null> {
	return fetchStandings<CurrentConstructorStandings>(
		1,
		`${ERGAST_API_ENDPOINT}/current/constructorstandings/?format=json`,
		(standings) => {
			let output = `🔧 \x02FORMULA 1 ${standings.MRData.StandingsTable.season} WCC Standings\x02:`;
			const standingsList = standings.MRData.StandingsTable.StandingsLists[0];
			if (!standingsList) {
				return `No constructor standings yet for the ${standings.MRData.StandingsTable.season} season.`;
			}

			for (const standing of standingsList.ConstructorStandings.slice(0, 10)) {
				output += ` ${standing.position}. ${standing.Constructor.name} - \x0303[${standing.points}]\x03`;
			}

			return output;
		},
	);
}

export async function returnWdcStandings(): Promise<string | null> {
	return fetchStandings<CurrentDriverStandings>(
		0,
		`${ERGAST_API_ENDPOINT}/current/driverstandings/?format=json`,
		(standings) => {
			let output = `🏆 \x02FORMULA 1 ${standings.MRData.StandingsTable.season} WDC Standings\x02:`;
			const standingsList = standings.MRData.StandingsTable.StandingsLists[0];
			if (!standingsList) {
				return `No driver standings yet for the ${standings.MRData.StandingsTable.season} season.`;
			}

			for (const standing of standingsList.DriverStandings.slice(0, 10)) {
				output += ` ${standing.position}. ${standing.Driver.code} - \x0303[${standing.points}]\x03`;
			}

			return output;
		},
	);
}

function findResultByCode<T extends { Driver: { code?: string } }>(
	results: T[] | undefined,
	code: string,
): T | undefined {
	return results?.find((result) => result.Driver.code?.toUpperCase() === code);
}

function parsePosition(position: string | undefined): number | null {
	if (!position) return null;
	const parsed = Number.parseInt(position, 10);
	return Number.isNaN(parsed) ? null : parsed;
}

export async function fetchHeadToHead(leftCodeArg: string, rightCodeArg: string): Promise<string> {
	const leftCode = leftCodeArg.toUpperCase();
	const rightCode = rightCodeArg.toUpperCase();

	const [standingsResponse, raceResultsResponse, qualifyingResultsResponse] = await Promise.all([
		fetchJson<CurrentDriverStandings>(
			`${ERGAST_API_ENDPOINT}/current/driverstandings/?format=json`,
		),
		fetchJson<RaceTableResponse<{ Results: SessionResultEntry[] }>>(
			`${ERGAST_API_ENDPOINT}/current/results/?format=json&limit=1000`,
		),
		fetchJson<RaceTableResponse<{ QualifyingResults: SessionResultEntry[] }>>(
			`${ERGAST_API_ENDPOINT}/current/qualifying/?format=json&limit=1000`,
		),
	]);

	const standingsTable = standingsResponse.MRData.StandingsTable;
	const season = standingsTable.season;
	const driverStandings = standingsTable.StandingsLists[0]?.DriverStandings ?? [];
	const races = raceResultsResponse.MRData.RaceTable.Races;
	const qualifyingSessions = qualifyingResultsResponse.MRData.RaceTable.Races;
	if (driverStandings.length === 0 && races.length === 0 && qualifyingSessions.length === 0) {
		return `No H2H data yet for the ${season} season.`;
	}

	const standingsByCode = new Map(
		driverStandings.map((standing) => [standing.Driver.code?.toUpperCase(), standing] as const),
	);
	const leftStanding = standingsByCode.get(leftCode);
	if (!leftStanding) {
		return `Unknown driver code: ${leftCode}.`;
	}

	const rightStanding = standingsByCode.get(rightCode);
	if (!rightStanding) {
		return `Unknown driver code: ${rightCode}.`;
	}

	const drivers: Array<{
		standing: (typeof driverStandings)[number];
		raceWins: number;
		qualiWins: number;
		podiums: number;
		bestFinish: number | null;
	}> = [
		{ standing: leftStanding, raceWins: 0, qualiWins: 0, podiums: 0, bestFinish: null },
		{ standing: rightStanding, raceWins: 0, qualiWins: 0, podiums: 0, bestFinish: null },
	];

	for (const race of races) {
		const positions: [number | null, number | null] = [
			parsePosition(findResultByCode(race.Results, leftCode)?.position),
			parsePosition(findResultByCode(race.Results, rightCode)?.position),
		];

		for (const [index, position] of positions.entries()) {
			if (position === null) continue;
			const driver = drivers[index];
			if (position <= 3) driver.podiums++;
			if (driver.bestFinish === null || position < driver.bestFinish) {
				driver.bestFinish = position;
			}
		}

		if (positions[0] !== null && positions[1] !== null) {
			if (positions[0] < positions[1]) drivers[0].raceWins++;
			if (positions[1] < positions[0]) drivers[1].raceWins++;
		}
	}

	for (const session of qualifyingSessions) {
		const positions: [number | null, number | null] = [
			parsePosition(findResultByCode(session.QualifyingResults, leftCode)?.position),
			parsePosition(findResultByCode(session.QualifyingResults, rightCode)?.position),
		];

		if (positions[0] !== null && positions[1] !== null) {
			if (positions[0] < positions[1]) drivers[0].qualiWins++;
			if (positions[1] < positions[0]) drivers[1].qualiWins++;
		}
	}

	return `⚔️ \x02H2H ${leftCode} vs ${rightCode}\x02 (${season}): Race \x0303${drivers[0].raceWins}-${drivers[1].raceWins}\x03 | Quali \x0303${drivers[0].qualiWins}-${drivers[1].qualiWins}\x03 | Points \x0303${drivers[0].standing.points}-${drivers[1].standing.points}\x03 | Podiums \x0303${drivers[0].podiums}-${drivers[1].podiums}\x03 | Best finish \x0303${drivers[0].bestFinish ? `P${drivers[0].bestFinish}` : "-"}/${drivers[1].bestFinish ? `P${drivers[1].bestFinish}` : "-"}\x03`;
}

export async function fetchNextEvent(): Promise<string | null> {
	console.log("Checking for upcoming events...");

	const nextEvent = await getNextEvent();
	if (!nextEvent) {
		console.log("No upcoming events found in database");
		return null;
	}

	const { meetingName, eventType, startTime } = nextEvent;
	const eventTime = startTime * 1000;
	const timeToEventMinutes = Math.floor((eventTime - Date.now()) / (1000 * 60));

	console.log(
		`Found upcoming event: ${meetingName}, type: ${eventType}, time: ${new Date(eventTime).toISOString()}`,
	);
	console.log(`Time to event: ${timeToEventMinutes} minutes`);

	if (timeToEventMinutes <= 0 || timeToEventMinutes > 5) {
		console.log(
			`Event not starting soon (${timeToEventMinutes} minutes away), no notification needed`,
		);
		return null;
	}

	const eventTypeName = await getEventTypeName(eventType);
	console.log(`Event starting soon! Preparing notification for ${meetingName}: ${eventTypeName}`);
	return `🏎️ \x02${meetingName}: ${eventTypeName}\x02 begins in 5 minutes.`;
}
